docs: improve how-to and ref info about integrating KLT with apps (#1785)
keptn/lifecycle-toolkit: docs/content/en/docs/concepts/architecture/keptn-apps/_index.md

---
title: KeptnApp and KeptnWorkload resources
linktitle: Keptn Applications and Keptn Workloads
description: How Keptn applications work
weight: 50
cascade:
---

## Keptn Workloads

A
[KeptnWorkload](../../../crd-ref/lifecycle/v1alpha3/#keptnworkload)
resource augments a Kubernetes
[Workload](https://kubernetes.io/docs/concepts/workloads/)
with the ability to handle extra phases.

> It can execute the pre- and post-deployment checks of a Workload
> and run pre- and post-deployment tasks.

Reviewer: To avoid confusion, I would suggest to always use the terms `evaluation` and `tasks`, instead of `check`,  to establish a clear relation to their respective custom resources.
```suggestion
It can execute the pre- and post-deployment evaluations of a Workload
and run pre- and post-deployment tasks.
```

In its state, it keeps track of the currently active `Workload Instances`,

> (`Pod`, `DaemonSet`, `StatefulSet`, and `ReplicaSet resources),

Reviewer: ```suggestion
(`Pod`, `DaemonSet`, `StatefulSet`, and `ReplicaSet` resources),
```

as well as the overall state of the Pre Deployment phase,

> which the scheduler can use to determine whether the deployment should proceed.

Reviewer: ```suggestion
which the scheduler can use to determine whether the pods belonging to a workload should proceed.
```

When it detects that the referenced object has reached its desired state
(e.g. all pods of a deployment are up and running),
it knows that a`PostDeploymentCheck` can be triggered.

The KeptnWorkload resources are created automatically
by the mutating webhook as soon as a pod for the workload

> (i.e. deployment, statefulset, daemonset, replicaset)

Reviewer: ```suggestion
(i.e. `Deployment`, `StatefulSet`, `DaemonSet`, `ReplicaSet`)
```

is about to be started.
The KeptnWorkloads are created automatically and without delay by the webhook.

## Keptn Applications

A [KeptnApp](../../../yaml-crd-ref/app.md)
resource combines multiple Kubernetes
[workloads](https://kubernetes.io/docs/concepts/workloads/)
into a single entity
that represent the application that is published.
Note that the Kubernetes documentation
often refers to workloads as applications,
but each workload actually corresponds to one version
of one deployable microservice,
not the amalgamation of multiple microservices
that typically comprise the released software.

Implementing Keptn applications provides the following benefits:

* Observability tools report on the deployment
  of all workloads together rather than individually.
* You can define pre-deployment evaluations and tasks
  that must all complete successfully
  before the scheduler creates the pods for any of the workloads.
* You can define post-deployment evaluations and tasks
  that run only after all the workloads have completed successfully.

The `KeptnApp` resources that are generated automatically
contain the identifications required to run the KLT observability features.

> The `spec.workloads.name` and a `spec.workloads.version` fields

Reviewer: The `spec.workloads.name/version` fields are generated automatically, based on what Workloads share the same `app.kubernetes.io/part-of` or `keptn.sh/app` label/annotations.
To manually define the pre and post evaluations/tasks, the fields `spec.[preDeploymentTasks|postDeploymentTasks|preDeploymentEvaluations|postDeploymentEvaluations]` need to be set manually

Author: Whoops!  I knew this.  And I didn't include a link to the docs that explains the annotations/labels here -- all a bit muddy!  I think I fixed it.

---
title: KeptnApp and KeptnWorkload resources
linktitle: Keptn Applications and Keptn Workloads
description: How Keptn applications work
weight: 50
cascade:
---

## Keptn Workloads

A
[KeptnWorkload](../../../crd-ref/lifecycle/v1alpha3/#keptnworkload)
resource augments a Kubernetes
[Workload](https://kubernetes.io/docs/concepts/workloads/)
with the ability to handle extra phases.
It can execute the pre- and post-deployment evaluations of a Workload
and run pre- and post-deployment tasks.

> In its state, it keeps track of the currently active `Workload Instances`,

Author: ```suggestion
In its state, it tracks the currently active `Workload Instances`,
```

(`Pod`, `DaemonSet`, `StatefulSet`, and `ReplicaSet` resources),
as well as the overall state of the Pre Deployment phase,

> which the scheduler can use to determine whether the pods belonging to a workload should proceed.

Reviewer 2: ```suggestion
which the scheduler can use to determine whether the pods belonging to a workload should be created and assigned to a node.
```

Author: done

When it detects that the referenced object has reached its desired state
(e.g. all pods of a deployment are up and running),
it knows that a`PostDeploymentCheck` can be triggered.

The KeptnWorkload resources are created automatically

> by the mutating webhook as soon as a pod for the workload

Reviewer 2: ```suggestion
by the mutating webhook as soon as you apply their manifest
```

Author: Oh, now I see -- too many comments for the same paragraph ;-)  Check what I have now -- I went for the passive "as soon as the workload manifest is applied" -- I'm thinking that the user doesn't always directly apply a manifest...

(i.e. `Deployment`, `StatefulSet`, `DaemonSet`, `ReplicaSet`)

> is about to be started.

Reviewer 2: ```suggestion

```

Author: But this leaves us with an incomplete sentence.  I tried a little rewording.  But actually, we list off the types of workloads in the preceding paragraph -- maybe I should delete them here?  That gives a cleaner sentence: "...when the pod for the workload starts."  Whaddayathink?

The KeptnWorkloads are created automatically and without delay by the webhook.

## Keptn Applications

A [KeptnApp](../../../yaml-crd-ref/app.md)
resource combines multiple Kubernetes
[workloads](https://kubernetes.io/docs/concepts/workloads/)
into a single entity
that represent the application that is published.
Note that the Kubernetes documentation
often refers to workloads as applications,
but each workload actually corresponds to one version
of one deployable microservice,

> not the amalgamation of multiple microservices

Reviewer 2: I love this sentence ❤️

that typically comprise the released software.

Implementing Keptn applications provides the following benefits:

* Observability tools report on the deployment
  of all workloads together rather than individually.
* You can define pre-deployment evaluations and tasks
  that must all complete successfully
  before the scheduler creates the pods for any of the workloads.
* You can define post-deployment evaluations and tasks
  that run only after all the workloads have completed successfully.

You control the content of a `KeptnApp` resource
with annotations or labels that are applied to each
[Workload](https://kubernetes.io/docs/concepts/workloads/)
([Deployments](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/),
[StatefulSets](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/),
[DaemonSets](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/),
and

> [ReplicaSets](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/):

Reviewer 2: I would remove replicset synce user do not direcly create them

Author: done
